### PR TITLE
KAFKA-17435:remove use.incremental.alter.configs

### PR DESCRIPTION
**Pull request by gongxuanzhang**
remove `use.incremental.alter.configs` and associated code

### CODE REVIEW

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java [hunk 1]
```diff
@@ -645,33 +617,23 @@ void incrementalAlterConfigs(Map<String, Config> topicConfigs) throws ExecutionE
         }
         log.trace("Syncing configs for {} topics.", configOps.size());
         AtomicReference<Boolean> encounteredError = new AtomicReference<>(false);
```
Comment (chia7712) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java]: we don't need `encounteredError` now, right

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java [hunk 2]
```diff
@@ -135,7 +134,6 @@ public MirrorSourceConnector() {
         this.replicationPolicy = replicationPolicy;
```
Comment (chia7712) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java]: please remove this constructor

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for this patch.

**Comment by gongxuanzhang** [comment]
@chia7712  thanks for your review, i update code.

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java [hunk 3]
```diff
@@ -644,34 +604,22 @@ void incrementalAlterConfigs(Map<String, Config> topicConfigs) throws ExecutionE
...
-                                encounteredError.set(true);
-                                useIncrementalAlterConfigs = false;
-                            } else if (config.useIncrementalAlterConfigs().equals(MirrorSourceConfig.REQUIRE_INCREMENTAL_ALTER_CONFIGS)
-                                    && e instanceof UnsupportedVersionException && !encounteredError.get()) {
-                                log.error("Failed to sync configs for topic {} on cluster {} with IncrementalAlterConfigs API", k.name(), sourceAndTarget.target(), e);
-                                encounteredError.set(true);
-                                context.raiseError(new ConnectException("use.incremental.alter.configs was set to \"required\", but the target cluster '"
-                                        + sourceAndTarget.target() + "' is not compatible with IncrementalAlterConfigs API", e));
-                            } else {
-                                log.warn("Could not alter configuration of topic {}.", k.name(), e);
```
Comment (chia7712) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:667]: please keep this `else` condition.

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for this patch. could you please add this change to `upgrade.html` (https://github.com/apache/kafka/blob/trunk/docs/upgrade.html#L22)

**Comment by gongxuanzhang** [comment]
@chia7712  thanks for your review,i updated code

File: docs/upgrade.html [hunk 4]
```diff
@@ -30,6 +30,9 @@ <h5><a id="upgrade_400_notable" href="#upgrade_400_notable">Notable changes in 4
             </li>
```
Comment (chia7712) [inline docs/upgrade.html]: Please add the docs for the removed configs explicitly.

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updates

**Comment by gongxuanzhang** [comment]
@chia7712  i updated `upgrade.html`  PTAL

File: docs/upgrade.html [hunk 5]
```diff
@@ -30,6 +30,12 @@ <h5><a id="upgrade_400_notable" href="#upgrade_400_notable">Notable changes in 4
...
         </ul>
     </ul>
+    <ul>
+        <li>A number of deprecated configurations have been removed, with some configurations fully deleted and others having outdated enums removed.</li>
```
Comment (chia7712) [inline docs/upgrade.html]: we don't need this duplicate description, right?

Comment (gongxuanzhang) [reply docs/upgrade.html]: The previous description has no `configuration`.
I think the`remove configuration description` will be a lot of code. So I wrote another section.
I removed it now, PTAL.

Comment (chia7712) [reply docs/upgrade.html]: I feel you can add "configutation" to it :)

Comment (gongxuanzhang) [reply docs/upgrade.html]: please check this version

**Comment by chia7712** [comment]
@gongxuanzhang please fix the conflicts

File: docs/upgrade.html [hunk 6]
```diff
@@ -33,6 +33,7 @@ <h5><a id="upgrade_400_notable" href="#upgrade_400_notable">Notable changes in 4
                 <a href="/{{version}}/javadoc/org/apache/kafka/tools/api/RecordReader.html"><code>org.apache.kafka.tools.api.RecordReader</code></a>
                 interface to build custom readers for the <code>kafka-console-producer</code> tool.
             </li>
+            <li> Remove <code>use.incremental.alter.configs</code>. The modified behavior is identical to the previous <code>required</code> configuration </li>
```
Comment (chia7712) [inline docs/upgrade.html]: Could you add "therefore users should ensure that target broker is at least 2.3.0" to the docs

**Review by chia7712 — COMMENTED** [review]
@gongxuanzhang thanks for updated PR

**Comment by gongxuanzhang** [comment]
@chia7712  thanks for your review ,PTAL

**Comment by chia7712** [comment]
@gongxuanzhang could you please fix the conflicts?

**Review by chia7712 — APPROVED** [review]

File: connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java [hunk 7]
```diff
@@ -644,34 +604,24 @@ void incrementalAlterConfigs(Map<String, Config> topicConfigs) throws ExecutionE
...
+                        log.error("Failed to sync configs for topic {} on cluster {} with " +
+                                "IncrementalAlterConfigs API", k.name(), sourceAndTarget.target(), e);
+                        context.raiseError(new ConnectException("the target cluster '"
+                                + sourceAndTarget.target() + "' is not compatible with " +
+                                "IncrementalAlterConfigs " +
+                                "API", e));
+                    } else {
+                        log.warn("Could not alter configuration of topic {}.", k.name(), e);
+                    }
+                }));
```
Comment (showuon) [inline connect/mirror/src/main/java/org/apache/kafka/connect/mirror/MirrorSourceConnector.java:620]: Unexpected WARN log will be printed even if the alterConfig completes without error (i.e. `e == null`). Opened [KAFKA-20017](https://issues.apache.org/jira/browse/KAFKA-20017) to fix it.

**Pull request merged**